firmware/src/boards/cynthion_d11/fpga_adv.c

```diff
@@ -0,0 +1,153 @@
+
+// Create a reference to our SERCOM object.
+typedef Sercom sercom_t;
+static sercom_t *sercom = SERCOM1;
+
+static void fpga_adv_byte_received_cb(uint8_t byte, int parity_error);
+
+#endif
+
+/**
+ * Initialize FPGA_ADV receive-only serial port
+ */
+void fpga_adv_init(void)
+{
+#ifdef BOARD_HAS_USB_SWITCH
+	gpio_set_pin_direction(FPGA_ADV, GPIO_DIRECTION_IN);
+	gpio_set_pin_pull_mode(FPGA_ADV, GPIO_PULL_UP);
+
+	// Disable the SERCOM before configuring it, to 1) ensure we're not transacting
+	// during configuration; and 2) as many of the registers are R/O when the SERCOM is enabled.
+	while(sercom->USART.SYNCBUSY.bit.ENABLE);
+	sercom->USART.CTRLA.bit.ENABLE = 0;
+
+	// Software reset the SERCOM to restore initial values.
+	while(sercom->USART.SYNCBUSY.bit.SWRST);
+	sercom->USART.CTRLA.bit.SWRST = 1;
+
+	// The SWRST bit becomes accessible again once the software reset is
+	// complete -- we'll use this to wait for the reset to be finshed.
+	while(sercom->USART.SYNCBUSY.bit.SWRST);
+
+	// Ensure we can work with the full SERCOM.
+	while(sercom->USART.SYNCBUSY.bit.SWRST || sercom->USART.SYNCBUSY.bit.ENABLE);
+
+	// Pinmux the relevant pins to be used for the SERCOM.
+	gpio_set_pin_function(PIN_PA09, MUX_PA09C_SERCOM1_PAD3);
+
+	// Set up clocking for the SERCOM peripheral.
+	_pm_enable_bus_clock(PM_BUS_APBC, SERCOM1);
+	_gclk_enable_channel(SERCOM1_GCLK_ID_CORE, GCLK_CLKCTRL_GEN_GCLK0_Val);
+
+	// Configure the SERCOM for UART mode.
+	sercom->USART.CTRLA.reg =
+		SERCOM_USART_CTRLA_DORD     |          // LSB first
+		SERCOM_USART_CTRLA_RXPO(3)  |          // RX on PA09 (PAD[3])
+		SERCOM_USART_CTRLA_SAMPR(0) |          // use 16x oversampling
+		SERCOM_USART_CTRLA_FORM(1)	|          // enable parity
+		SERCOM_USART_CTRLA_RUNSTDBY |          // don't autosuspend the clock
+		SERCOM_USART_CTRLA_MODE_USART_INT_CLK; // use internal clock
+
+	// Configure our baud divisor.
+	const uint32_t baudrate = 9600;
+	const uint32_t baud = (((uint64_t)CONF_CPU_FREQUENCY << 16) - ((uint64_t)baudrate << 20)) / CONF_CPU_FREQUENCY;
+	sercom->USART.BAUD.reg = baud;
+
+	// Configure TX/RX and framing.
+	sercom->USART.CTRLB.reg =
+			SERCOM_USART_CTRLB_CHSIZE(0) | // 8-bit words
+			SERCOM_USART_CTRLB_RXEN;       // Enable RX.
+
+	// Wait for our changes to apply.
+	while (sercom->USART.SYNCBUSY.bit.CTRLB);
+
+	// Enable our receive interrupt, as we want to asynchronously dump data into
+	// the UART console.
+	sercom->USART.INTENSET.reg = SERCOM_USART_INTENSET_RXC;
+
+	// Enable the UART IRQ.
+	NVIC_EnableIRQ(SERCOM1_IRQn);
+
+	// Finally, enable the SERCOM.
+	sercom->USART.CTRLA.bit.ENABLE = 1;
+	while(sercom->USART.SYNCBUSY.bit.ENABLE);
+
+	// Update timestamp
+	last_phy_adv = board_millis();
+#endif
+}
+
+/**
+ * Task for things related with the advertisement pin
+ */
+void fpga_adv_task(void)
+{
+#ifdef BOARD_HAS_USB_SWITCH
+	// Take over USB after timeout
+	if (board_millis() - last_phy_adv < TIMEOUT) return;
+	take_over_usb();
```

Comment:
I think this would be a bit clearer as:

```suggestion
        // Take over USB after timeout
	if (board_millis() - last_phy_adv >= TIMEOUT) {
		take_over_usb();
	}
```
```